cmd/labs/project/project.go

```diff
@@ -0,0 +1,294 @@
+	for _, v := range dirs {
+		err := os.MkdirAll(v, 0o744)
+		if err != nil {
+			return fmt.Errorf("folder %s: %w", v, err)
+		}
+	}
+	return nil
+}
+
+func (p *Project) virtualEnvPath(ctx context.Context) string {
+	activeVenv := env.Get(ctx, "VIRTUAL_ENV")
+	if p.IsDeveloperMode(ctx) && activeVenv != "" {
+		// When a virtual environment has been activated, the VIRTUAL_ENV environment variable
+		// is set to the path of the environment. Since explicitly activating a virtual environment
+		// is not required to use it, VIRTUAL_ENV cannot be relied upon to determine whether a virtual
+		// environment is being used.
+		//
+		// See https://docs.python.org/3/library/venv.html#how-venvs-work
+		return activeVenv
+	}
+	return filepath.Join(p.StateDir(ctx), "venv")
+}
+
+func (p *Project) virtualEnvPython(ctx context.Context) string {
+	pythonExecutable := "python3"
+	if runtime.GOOS == "windows" {
+		pythonExecutable = "python3.exe"
+	}
+	return filepath.Join(p.virtualEnvPath(ctx), "bin", pythonExecutable)
+}
+
+func tryLoadAndParseJSON[T any](jsonFile string) (*T, error) {
+	raw, err := os.ReadFile(jsonFile)
+	if errors.Is(err, fs.ErrNotExist) {
+		return nil, err
+	}
+	if err != nil {
+		return nil, fmt.Errorf("read %s: %w", jsonFile, err)
+	}
+	var v T
+	err = json.Unmarshal(raw, &v)
+	if err != nil {
+		return nil, fmt.Errorf("parse %s: %w", jsonFile, err)
+	}
+	return &v, nil
+}
+
+func (p *Project) authFile(ctx context.Context) string {
+	return filepath.Join(p.ConfigDir(ctx), "auth.json")
+}
+
+func (p *Project) loadAuthConfig(ctx context.Context) (*authConfig, error) {
+	authFile := p.authFile(ctx)
+	log.Debugf(ctx, "Loading auth configuration from: %s", authFile)
+	return tryLoadAndParseJSON[authConfig](authFile)
+}
+
+func (p *Project) writeAuthConfig(ctx context.Context, ac authConfig) error {
+	authFile := p.authFile(ctx)
+	raw, err := json.Marshal(ac)
+	if err != nil {
+		return err
+	}
+	log.Debugf(ctx, "Writing auth configuration to: %s", authFile)
+	return os.WriteFile(authFile, raw, 0o600)
+}
+
+func (p *Project) versionFile(ctx context.Context) string {
+	return filepath.Join(p.StateDir(ctx), "version.json")
+}
+
+func (p *Project) InstalledVersion(ctx context.Context) (*version, error) {
+	versionFile := p.versionFile(ctx)
+	log.Debugf(ctx, "Loading installed version info from: %s", versionFile)
+	return tryLoadAndParseJSON[version](versionFile)
+}
+
+func (p *Project) writeVersionFile(ctx context.Context, ver string) error {
+	versionFile := p.versionFile(ctx)
+	raw, err := json.Marshal(version{
+		Version: ver,
+		Date:    time.Now(),
+	})
+	if err != nil {
+		return err
+	}
+	log.Debugf(ctx, "Writing installed version info to: %s", versionFile)
+	return os.WriteFile(versionFile, raw, 0o600)
+}
+
+func (p *Project) checkUpdates(cmd *cobra.Command, _ []string) error {
+	ctx := cmd.Context()
+	r := releases.NewReleaseCache("databrickslabs", p.Name, p.CacheDir(ctx))
```

Comment:
How does this cache dir related to the one used in the installer?

Reply:
it doesn't. but `databricks labs isntall X` doesn't know the version of X, so it has to have a separate path to access the cache. `checkUpdates` works with already installed project